Adding improvements to iceberg on cloud tests

rpdevmp opened the pull request:
## Backports Required

- [x ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x

## Release Notes
* none

### Improvements
This PR add improvements to Iceberg on cloud tests:

1. table_verifier.py that could be used by different catalogs and query engines. Comparing SQL tables, values and headers.
2. Force cleanup of Databricks Unity external locations (as suggested by Databricks team)
3. Adding my email to one of the principals on DBX. Admin user should be able to see created resources (and used to be able to see and manage those), starting recently admin user cannot see created items by another principal.. Reported to DBX team. Adding my email as a workaround to be able to debug in DBX portal SQL editor if needed
4.  Producing data key_value, as text, json and json with headers and added verifications of tables
5. Added additional test to measure cloud config updates (this would not be part of regular test suite, but was used multiple times to measure cloud config update performance and to try reproduce an older bug)

P.S.
This was run many times on local branch with Ducktape infra and BYOC clusters

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--802301c1-caba-4892-aece-9b44e9c40090-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/69732">build#69732</a></summary>


| test_class                     | test_method                       | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:-------------------------------|:----------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| CompactionTermRollRecoveryTest | test_compact_term_rolled_recovery | null             | integration | https://buildkite.com/redpanda/redpanda/builds/69732#01984f86-f444-46fc-a6ea-962422bee6f3 | FLAKY         | 20/21    | upstream reliability is '99.70501474926253'. current run reliability is '95.23809523809523'. drift is 4.46692 and the allowed drift is set to 50. The test should PASS |
</details>
<!--/body-->

nvartolomei commented on tests/rptest/redpanda_cloud_tests/iceberg_catalogs_test.py:
```diff
@@ -249,20 +251,91 @@ def test_databricks_basic(self):
+            # Create topic(s) and produce data
+            self.rpk = RpkTool(self.redpanda)
+            self.test_topic = f"test_topic_{random.randint(10000, 99999)}"
+            self.logger.debug(f"Creating Iceberg topic: self.test_topic")
```
bad f-string formatting. missing braces around variable

rpdevmp replied:
Good catch

nvartolomei commented on tests/rptest/redpanda_cloud_tests/iceberg_catalogs_test.py:
```diff
@@ -249,20 +251,91 @@ def test_databricks_basic(self):
+                self.logger.info(f"...waiting {i} seconds remaining")
+            time.sleep(1)
+
+        expected_records = [
```
where did the `foo {i} bar {i}` records go produced on line 267?

rpdevmp replied:
Same as other produced data this goes into self.test_topic, it creates a topic and then produce records there. Topic names have random names, so that test could be re-run on the same BYOC cluster and create fresh topic

nvartolomei commented on tests/rptest/services/databricks_workspace.py:
```diff
@@ -62,6 +63,10 @@ def create_catalog(self, bucket: str) -> "DatabricksCatalogInfo":
         self._location_names.add(bucket)
 
+        # Clean up any previous resources if reusing bucket name
+        self._cleanup_catalogs()
```
This does not have any effect. `_cleanup_*` methods work on resources which are tracked by the service (see self._catalog_names for example). They can't cleanup anything which was not created by this service. This is on purpose. I.e. to avoid stepping over other concurrently running tests.

This periodic cleanup must be done outside of ducktape imho. The service/each test shouldn't spend time inspecting entire account to figure out what to clean up. Waste of time.

nvartolomei commented on tests/rptest/services/databricks_workspace.py:
```diff
@@ -115,19 +120,22 @@ def create_catalog(self, bucket: str) -> "DatabricksCatalogInfo":
-            cursor.execute(sql)
-            self.logger.debug("Granted PRIVILEGES successfully")
+            # List of principals to grant permissions to (adding users for debugging)
+            principals = [principal, "marat.pekker@redpanda.com"]
```
I don't believe this is necessary. When you want to debug something you can grant yourself privileges from the databricks workspace when you need to imho. This doesn't belong to the test.

rpdevmp replied:
That is how it used to work.. Lately, it did not allow me to do that even with admin right, but I agree that better to to add it to the test, will remove

nvartolomei commented on tests/rptest/services/datalake/query_engine/databricks_sql.py:
```diff
@@ -59,3 +60,39 @@ def optimize_parquet_files(self, namespace: str, table: str) -> None:
             f"OPTIMIZE {self.escape_identifier(namespace)}.{self.escape_identifier(table)}"
         )
         self.logger.debug(f"OPTIMIZE result: {result}")
+        raise NotImplementedError(
```
This is implemented and shouldn't raise. Moreover, it is used successfully in one test https://github.com/redpanda-data/redpanda/blob/3a4d5a21a5f50434a84e26dae872bc8cc9011517/tests/rptest/tests/datalake/databricks_test.py#L403

rpdevmp replied:
This was not supposed to be part of this PR

nvartolomei commented on tests/rptest/services/datalake/query_engine/databricks_sql.py:
```diff
@@ -59,3 +60,39 @@ def optimize_parquet_files(self, namespace: str, table: str) -> None:
+        raise NotImplementedError(
+            "DatabricksSQL optimize_parquet_files is not implemented yet")
+
+    def verify_table_contents(
```
This method does not belong to the databricks_sql service. The verifier is separate functionality which depends on a query engine/maybe databricks specific query engine.

Let's separate concerns: client/sql service vs verifier.

nvartolomei replied:
You also might be interested in looking at tests/rptest/tests/datalake/datalake_verifier.py.

nvartolomei commented on tests/rptest/services/datalake/util/table_verifier.py:
```diff
@@ -0,0 +1,87 @@
+
+
+def canonical_json(val):
+    """Serialize dicts to canonical JSON strings for comparison."""
```
This method doesn't do what the name and comment says. It returns strings only for dicts and doesn't do anything for all the other input types.

rpdevmp committed:
datalake data verification and other improvements

rpdevmp commented:
@nvartolomei 
- Updated datalake verifier with actual data verification. It will check the topic tables, check for matching data, verify actual fields
- You should be able to also use this for non-cloud CDT verifications for core tests. Specify topic and provide expected data
- external location cleanup fix
- tested data verification and re-consolation tests on cloud pipelines
- added logic for other cloud providers when we add storage object URL

rpdevmp committed:
datalake data verification and other improvements

rpdevmp committed:
Merge branch 'dev' into rpdevmp/iceberg-cloud-improvements

nvartolomei commented on tests/rptest/tests/datalake/datalake_verifier.py:
```diff
@@ -227,6 +241,25 @@ def _get_query(self, partition, last_queried_offset, max_consumed_offset):
+        # 1. A struct column `redpanda` containing partition, offset, timestamp, headers, key
+        # 2. A `value` column containing the message payload (hex encoded)
+        # Note: For Databricks, we need the full catalog.namespace.table format
+        if hasattr(self._query, '_catalog_name'):
```
This is very hacky way of checking for a catalog type.

And beside, this is not needed. We have queries in other places like the verifier, the count query QueryEngineBase which work fine with catalog name only across all query engines including databricks.

nvartolomei commented on tests/rptest/tests/datalake/datalake_verifier.py:
```diff
@@ -227,6 +241,25 @@ def _get_query(self, partition, last_queried_offset, max_consumed_offset):
         AND redpanda.offset<={max_consumed_offset} \
         ORDER BY redpanda.offset"
 
+    def _get_query_data(self):
```
Why don't we extend `_get_query` and `_verify_next_message` to also do data verification? Probably this also needs customization to handle the key_value vs schemafull modes. I'm fine with handling only key_value for now if no tests fail.

nvartolomei replied:
Also no need to buffer anything/everything in memory. Early stopping on errors.

nvartolomei commented on tests/rptest/tests/datalake/datalake_verifier.py:
```diff
@@ -411,3 +463,227 @@ def oneshot(
+        if isinstance(val, memoryview):
+            return val.tobytes().decode("utf-8", errors="replace")
+        # Check if it's a hex string (from Databricks)
+        if isinstance(val, str) and len(val) > 0 and all(
```
This is odd choice of decoding binary values. Why we need to jump with all these hoops with bytes/memoryview/hex decoding?

Maybe we need to push some responsibility into query engine so it tells us how to access raw binary?

nvartolomei replied:
I.e. maybe we need to apply a function in the query, i.e. `encodeHex()` and we have a method on all query engines which they give their own hex encoding method and the on the python side we always do hex decode and expect it to succeed. Or, we have a `decodeBinary` which each queryengine implements?

`replace` error handler on utf-8 decoding is also odd. Why not strict? Do we expect these bytes to be utf-8 valid at all? I don't understand this code.

nvartolomei reviewed (commented):
This needs to be reformatted properly.

It needs to be split into separate commits, i.e. changing existing things, adding new functionality, updating the test all likely should be separate commits with messages explaining why these changes are introduced, what was wrong before, how we are fixing it and why.

The additions to datalake verifier needs extra attention. The idea seems fine but we need a clean implementation so that it is intelligible, maintainable, extendable. In the current form it has none of these qualities.
